File: ryh/com/leetcode/LeetCode_test.java
package com.leetcode;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.HashMap;
import java.util.Map;
public class LeetCode_test {
	//7：反转整数
	int reverse(int x){
		int rev = 0;
		while(x!=0){
			int pop = x % 10;
			x/=10;
			if(rev > Integer.MAX_VALUE/10 || (rev == Integer.MAX_VALUE/10 && pop>7))
			{
				return 0;
			}
			if(rev <Integer.MIN_VALUE/10 ||(rev == Integer.MIN_VALUE/10 && pop<-8))
			{
				return 0;
			}
			rev = rev*10+pop;
		}
		return rev;
	}
	//判断是否是回文数
	boolean isPalindrome(int x){
		if(x==0){
			return true;
		}
		else if(x < 0){
			return false;
		}
		else{
			int result = 0;

			int y = x;

			while(y!=0){
				int p = y%10;
				y = y/10;

				result = result * 10 + p;
			}
			return result==x;
		}
	}
	//罗马文字转数字
	int romanToInt(String s)
	{
		Map<Character,Integer> map = new HashMap();
		map.put('I',1);
		map.put('V',5);
		map.put('X',10);
		map.put('L',50);
		map.put('C',100);
		map.put('D',500);
		map.put('M',1000);

		int result = 0;
		for(int i=0;i<s.length();i++)
		{
			//
			if(i>=1 && map.get(s.charAt(i))>map.get(s.charAt(i-1)))
			{
				result += map.get(s.charAt(i))- 2*map.get(s.charAt(i-1));
			}
			else
			{
				result += map.get(s.charAt(i));
			}
		}
		return result;
	}
	//最长公共前缀
	public String longestCommonPreix(String [] strs) {

		if (strs.length == 0) {
			return "";
		}
		String prefix = strs[0];
		for (int i = 1; i < strs.length; i++) {
			while(strs[i].indexOf(prefix)!=0){
				prefix = prefix.substring(0,prefix.length()-1);
			}
			if(prefix.isEmpty()){
				return "";
			}
		}
		return prefix;
	}
	//是否是有效的括号
	boolean isValid(String s) {
		Deque<Character> stack = new ArrayDeque<>();
		for(int i=0;i<s.length();i++)
		{
			if(s.charAt(i) == '{' || s.charAt(i) == '[' || s.charAt(i) == '(')
			{
				stack.push(s.charAt(i));
			}
			else
			{
				if(stack.isEmpty())
					return false;
				else
				{
					if(stack.peek()=='(' && s.charAt(i)!=')')
						return false;
					else if(stack.peek()=='[' && s.charAt(i)!=']')
						return false;
					else if(stack.peek()=='{' && s.charAt(i)!='}')
						return false;
					stack.pop();
				}
			}
		}
		return stack.isEmpty();
	}
}

File: ryh/com/leetcode/LeetCode_test.php
<?php
//9：回文数
function isPalindrome($x) {
    if($x<0){
        return false;
    }
    if ($x == 0){
        return true;
    }
    $result = 0;
    $y = $x;
    while($y!=0){
        $p = $y%10;
        $y = intval($y/10);
        $result = $result * 10 + $p;
    }
    if( $x==$result){
        return true;
    }
    else {
        return false;
    }
}
//13：罗马字母转数字
function romanToInt($s) {
    $arr = array(
        'I'=>1,
        'V'=>5,
        'X'=>10,
        'L'=>50,
        'C'=>100,
        'D'=>500,
        'M'=>1000,
    );
    $tmp[] = str_split($s);
    $result = 0;
    for($i=0; $i<count($tmp[0]);$i++)
    {
        if($i>=1 && $arr[$tmp[0][$i]]>$arr[$tmp[0][$i-1]])
        {
            $result += $arr[$tmp[0][$i]]-2*$arr[$tmp[0][$i-1]];
        }
        else
        {
            $result += $arr[$tmp[0][$i]];
        }
    }
    return $result;
}
//14：最长公共前缀
function longestCommonPrefix($strs)
{
    if(count($strs)==1)
        return $strs[0];
    if(count($strs)==0)
        return "";
    $prefix = $strs[0];
    for($i=0;$i<count($strs);$i++)
    {
        $str = "";
        for($j=0;$j<min(strlen($prefix),strlen($strs[$i]));$j++)
        {
            if($prefix[$j] == $strs[$i][$j])
                $str.= $strs[$i][$j];
            else
                break;
        }
        $prefix = $str;
        if(empty($prefix))
        {
            break;
        }
    }
    return $prefix;
}
//判断是否是有效的括号
function isValid($s) {
    $arr = array();
    for($i=0;$i<strlen($s);$i++)
    {
        if($s[$i] == '(' || $s[$i] == '[' || $s[$i] == '{'){
            $arr[$i] = $s[$i];
        }
        else{
            if(count($arr) == 0){
                return false;
            }
            if($arr[count($arr)-1] == '(' && $s[$i] !=')')
            {
                return false;
            }
            if($arr[count($arr)-1] == '[' && $s[$i] !=']')
            {
                return false;
            }
            if($arr[count($arr)-1] == '{' && $s[$i] !='}')
            {
                return false;
            }
            unset($arr[count($arr)-1]);
        }
    }
    if(empty($arr))
    {
        return true;
    }
}
